libr/bin/p/bin_xtr_sep64.c

```diff
@@ -0,0 +1,493 @@
+	ut64 total_size;
+} RSepSlice64;
+
+typedef struct _RSepXtr64Ctx {
+	RSepHdr64 * hdr;
+	RSepApp64 * apps;
+} RSepXtr64Ctx;
+
+static RSepXtr64Ctx * sep64_xtr_ctx_new(RBuffer * buf);
+static void sep64_xtr_ctx_free(void * p);
+static RSepSlice64 * sep64_xtr_ctx_get_slice(RSepXtr64Ctx * ctx, RBuffer * whole, int idx);
+
+static RSepMachoInfo * mach0_info_new(RBuffer * buf, ut64 at, ut64 max_size);
+static void mach0_info_free(RSepMachoInfo * info);
+
+static ut32 read_arm64_ins(RBuffer *b, int idx);
+static char * get_proper_name (const char * app_name);
+static RBuffer * extract_slice(RBuffer * whole, RSepMachoInfo * info);
+static inline void fill_metadata_info_from_hdr(RBinXtrMetadata *meta, struct MACH0_(mach_header) *hdr);
+
+#define BTW(val, min, max) ((val) > min && (val) < max)
+
+static bool check_buffer(RBuffer *b) {
+	r_return_val_if_fail (b, false);
+
+	const ut64 sz = r_buf_size (b);
+	if (sz < 0x11c0) {
+		return false;
+	}
+
+	ut32 msr_vbar_el1 = read_arm64_ins (b, 2);
+	if (msr_vbar_el1 != 0xd518c002) {
+		return false;
+	}
+	ut32 adr = read_arm64_ins (b, 1);
+	if (adr != 0x10003fe2) {
+		return false;
+	}
+
+	/* check exception vector */
+	if (read_arm64_ins (b, 512) != 0x14000000) {
+		return false;
+	}
+	if (read_arm64_ins (b, 1023) != 0x14000000) {
+		return false;
+	}
+
+	/* legion2 */
+	if (read_arm64_ins (b, 1028) !=  0x326e6f69) {
+		return false;
+	}
+
+	/* data header start */
+	ut64 hdr_offset = read_arm64_ins (b, 1029);
+	if (hdr_offset >= sz) {
+		return false;
+	}
+
+	/* check size */
+	if (r_buf_read_le64_at (b, hdr_offset + 56) != sz) {
+		return false;
+	}
+
+	return true;
+}
+
+static bool load(RBin *bin) {
+	return ((bin->cur->xtr_obj = sep64_xtr_ctx_new (bin->cur->buf)) != NULL);
+}
+
+static void destroy(RBin *bin) {
+	sep64_xtr_ctx_free (bin->cur->xtr_obj);
+}
+
+static int size(RBin *bin) {
+	// TODO
+	return 0;
+}
+
+static RBinXtrData *oneshot_buffer(RBin *bin, RBuffer *b, int idx) {
+	r_return_val_if_fail (bin && bin->cur, NULL);
+
+	if (!bin->cur->xtr_obj) {
+		bin->cur->xtr_obj = sep64_xtr_ctx_new (b);
+	}
+	RSepXtr64Ctx * ctx = bin->cur->xtr_obj;
+
+	RSepSlice64 * slice = sep64_xtr_ctx_get_slice (ctx, b, idx);
+	RBinXtrData * res = r_bin_xtrdata_new (slice->buf, slice->nominal_offset, slice->total_size, 3 + ctx->hdr->n_apps, slice->meta);
+
+	r_buf_free (slice->buf);
+	free (slice);
+	return res;
+}
+
+static RList * oneshotall_buffer(RBin *bin, RBuffer *b) {
+	RBinXtrData *data = oneshot_buffer (bin, b, 0);
+	if (data) {
+		int narch = data->file_count;
+		RList *res = r_list_newf (r_bin_xtrdata_free);
+		if (!res) {
+			r_bin_xtrdata_free (data);
+			return NULL;
+		}
+		r_list_append (res, data);
+		int i;
+		for (i = 1; data && i < narch; i++) {
+			data = oneshot_buffer (bin, b, i);
+			r_list_append (res, data);
+		}
+		return res;
+	}
+	return NULL;
+}
+
+static RSepXtr64Ctx * sep64_xtr_ctx_new(RBuffer * buf) {
+	RSepHdr64 * hdr = NULL;
+	RSepApp64 * apps = NULL;
+	RSepXtr64Ctx * ctx = NULL;
```

Comment:
Define and assign in the same line below

Reply:
hmm moving the declarations down will complicate the `goto beach;` "pattern"